src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -768,8 +765,8 @@ public static void GetInfoForImmediateNamespaceMembers(
             int namespaceNameLength,
             IEnumerable<IGrouping<string, TypeDefinitionHandle>> typesByNS,
             StringComparer nameComparer,
-            out IEnumerable<IGrouping<string, TypeDefinitionHandle>> types,
-            out IEnumerable<KeyValuePair<string, IEnumerable<IGrouping<string, TypeDefinitionHandle>>>> namespaces)
+            [NotNull] out IEnumerable<IGrouping<string, TypeDefinitionHandle>>? types,
+            [NotNull] out IEnumerable<KeyValuePair<string, IEnumerable<IGrouping<string, TypeDefinitionHandle>>>>? namespaces)
```

Comment:
Perhaps we could simplify this to:
```suggestion
            out IEnumerable<IGrouping<string, TypeDefinitionHandle>> types,
            out IEnumerable<KeyValuePair<string, IEnumerable<IGrouping<string, TypeDefinitionHandle>>>> namespaces)
```

and move the asserts at the end of this method to before the assignments, something like this:
```cs
Debug.Assert(nestedTypes != null);
types = nestedTypes;
```
```cs
Debug.Assert(nestedNamespaces != null);
namespaces = nestedNamespaces;
```

Reply:
This is a signature I struggled with a bit but after reading I believed that the `[NotNull]` was the correct way to represent this. It seemed to follow from the other use of `out` in `Try` style APIs where we use `[NotNullWhen]`.

Will defer to @RikkiGibson, @jcouv if they feel differently.

Reply:
https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/api-guidelines/nullability.md#annotation-guidance

> DO annotate ref arguments that guarantee the argument will be non-null upon exit (e.g. lazy initialization helpers) with [NotNull].

This seems the advised approach.